### PR TITLE
[Plugin] Implement custom tool API

### PR DESCRIPTION
**Pull request by IntelOrca**
Introduces a new API for developing custom tools. Also includes an API for controlling the selection of map tiles.

**Example:**
```js
/// <reference path="C:/Users/Ted/source/repos/openrct2/distribution/openrct2.d.ts" />

var downCoord;
var currentCoord;

function selectTheMap() {
    var left = Math.min(downCoord.x, currentCoord.x);
    var right = Math.max(downCoord.x, currentCoord.x);
    var top = Math.min(downCoord.y, currentCoord.y);
    var bottom = Math.max(downCoord.y, currentCoord.y);
    ui.tileSelection.range = {
        leftTop: { x: left, y: top },
        rightBottom: { x: right, y: bottom }
    };
}

function getTileSurfaceZ(x, y) {
    var tile = map.getTile(x / 32, y / 32);
    if (tile) {
        for (var i = 0; i < tile.numElements; i++) {
            var element = tile.getElement(i);
            if (element && element.type == "surface") {
                return element.baseHeight * 8;
            }
        }
    }
    return null;
}

function fillTheMap() {
    var left = Math.min(downCoord.x, currentCoord.x);
    var right = Math.max(downCoord.x, currentCoord.x);
    var top = Math.min(downCoord.y, currentCoord.y);
    var bottom = Math.max(downCoord.y, currentCoord.y);
    for (var x = left; x <= right; x += 32) {
        for (var y = top; y <= bottom; y += 32) {
            var z = getTileSurfaceZ(x, y);
            if (z !== null) {
                context.executeAction("smallsceneryplace", {
                    x: x,
                    y: y,
                    z: z,
                    direction: 0,
                    quadrant: 0,
                    object: 169,
                    primaryColour: 0,
                    secondaryColour: 0
                });
            }
        }
    }
}

var main = function () {
    if (typeof ui === 'undefined') {
        return;
    }

    ui.registerMenuItem("Place Scenery Block", function() {
        if (ui.tool && ui.tool.id == "place-sc-tool") {
            ui.tool.cancel();
        } else {
            ui.activateTool({
                id: "place-sc-tool",
                cursor: "cross_hair",
                onStart: function(e) {
                    ui.mainViewport.visibilityFlags |= (1 << 7);
                },
                onDown: function(e) {
                    downCoord = e.mapCoords;
                    currentCoord = e.mapCoords;
                },
                onMove: function(e) {
                    if (e.isDown) {
                        currentCoord = e.mapCoords;
                        selectTheMap();
                    } else {
                        downCoord = e.mapCoords;
                        currentCoord = e.mapCoords;
                        selectTheMap();
                    }
                },
                onUp: function(e) {
                    fillTheMap();
                    ui.tileSelection.range = null;
                },
                onFinish: function() {
                    ui.tileSelection.range = null;
                    ui.mainViewport.visibilityFlags &= ~(1 << 7);
                },
            });
        }
    });
};

registerPlugin({
    name: 'Place Scenery Block',
    version: '1.0',
    authors: ['Ted'],
    type: 'local',
    main: main
});
```

### CODE REVIEW

File: distribution/openrct2.d.ts [hunk 1]
```diff
@@ -71,6 +71,14 @@ declare global {
         z: number;
     }
 
+    /**
+     * A rectangular area specified using two coordinates.
+     */
+    interface MapRange {
+        leftTop: Coord2;
+        rightBottom: Coord2;
```
Comment (tupaschoal) [inline distribution/openrct2.d.ts:79]: Looks like a good opportunity to give Coord2 more meaningful names, it's becoming very used.

File: distribution/openrct2.d.ts [hunk 2]
```diff
@@ -710,16 +718,89 @@ declare global {
...
+        readonly entityId?: number;
+    }
+
+    /**
+     * Describes the properties and event handlers for a custom tool.
+     */
+    interface ToolDesc {
+        id: string;
+        type: ToolType;
+        cursor: CursorType;
```
Comment (tupaschoal) [inline distribution/openrct2.d.ts:766]: Should this be optional and just allow the `arrow` to be the default?

File: distribution/openrct2.d.ts [hunk 3]
```diff
@@ -710,16 +718,89 @@ declare global {
...
+        "path_down" |
+        "dig_down" |
+        "water_down" |
+        "house_down" |
+        "volcano_down" |
+        "walk_down" |
+        "paint_down" |
+        "entrance_down" |
+        "hand_open" |
+        "hand_closed";
```
Comment (tupaschoal) [inline distribution/openrct2.d.ts:802]: This will be a lot more useful if sorted.

File: src/openrct2-ui/scripting/CustomMenu.cpp [hunk 4]
```diff
@@ -33,8 +46,169 @@ namespace OpenRCT2::Scripting
...
+
+    static DukValue CoordsXYToDuk(duk_context* ctx, const CoordsXY& coords)
+    {
+        DukObject dukCoords(ctx);
+        dukCoords.Set("x", coords.x);
+        dukCoords.Set("y", coords.y);
+        return dukCoords.Take();
+    }
+
+    static DukValue CoordsXYToDuk(duk_context* ctx, const ScreenCoordsXY& coords)
```
Comment (tupaschoal) [inline src/openrct2-ui/scripting/CustomMenu.cpp:94]: I'd either call these two `ToDuk` or specify that the second one is `ScreenCoordsXYToDuk`

File: src/openrct2-ui/scripting/CustomMenu.cpp [hunk 5]
```diff
@@ -33,8 +46,169 @@ namespace OpenRCT2::Scripting
...
+            duk_error(scriptEngine.GetContext(), DUK_ERR_ERROR, "Invalid parameters.");
+        }
+    }
+
+    static constexpr const char* CursorNames[] = {
+        "arrow",         "blank",      "up_arrow",      "up_down_arrow", "hand_point", "zzz",         "diagonal_arrows",
+        "picker",        "tree_down",  "fountain_down", "statue_down",   "bench_down", "cross_hair",  "bin_down",
+        "lamppost_down", "fence_down", "flower_down",   "path_down",     "dig_down",   "water_down",  "house_down",
+        "volcano_down",  "walk_down",  "paint_down",    "entrance_down", "hand_open",  "hand_closed",
+    };
```
Comment (tupaschoal) [inline src/openrct2-ui/scripting/CustomMenu.cpp:191]: This probably will fail `clang-format`, but in any case there's no point in creating a plain array, use an `std::array` or a `vector`

File: src/openrct2-ui/scripting/ScTileSelection.hpp [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+                auto dukX = dukCoords["x"];
+                if (dukX.type() == DukValue::Type::NUMBER)
+                {
+                    auto dukY = dukCoords["y"];
+                    if (dukY.type() == DukValue::Type::NUMBER)
+                    {
+                        CoordsXY coords;
+                        coords.x = dukX.as_int();
+                        coords.y = dukY.as_int();
+                        result = coords;
```
Comment (tupaschoal) [inline src/openrct2-ui/scripting/ScTileSelection.hpp:154]:
```suggestion
                        result = { dukX.as_int(), dukY.as_int() };
```

File: src/openrct2-ui/scripting/CustomMenu.cpp (hunk 5 above)
Comment (IntelOrca) [reply src/openrct2-ui/scripting/CustomMenu.cpp:191]: This is the work of `clang-format`.

**Comment by IntelOrca** [comment]
@tupaschoal Applied all your suggestions. In particular I have now defined generic template methods for converting to and from duk objects and our built-in types. The default implementation is deleted so each type must define their own implementation.

**Pull request merged**